XS test that needs investigation:

2020-04-07 16:45:40,456 - CRITICAL - EXCEPTION: test_03_list_snapshots: ['Traceback (most recent call last):\n', '  File "/usr/lib/python2.7/site-packages/nose/suite.py", line 228, in run\n    self.tearDown()\n', '  File "/usr/lib/python2.7/site-packages/nose/suite.py", line 351, in tearDown\n    self.teardownContext(ancestor)\n', '  File "/usr/lib/python2.7/site-packages/nose/suite.py", line 367, in teardownContext\n    try_run(context, names)\n', '  File "/usr/lib/python2.7/site-packages/nose/util.py", line 471, in try_run\n    return func()\n', '  File "/marvin/tests/smoke/test_list_ids_parameter.py", line 209, in tearDownClass\n    raise Exception("Warning: Exception during cleanup : %s" % e)\n', "Exception: Warning: Exception during cleanup : Job failed: {jobprocstatus : 0, created : u'2020-04-07T16:46:13+0000', jobresult : {errorcode : 530, errortext : u**'Failed to delete snapshot'**}, cmd : u'org.apache.cloudstack.api.command.user.snapshot.DeleteSnapshotCmd', userid : u'964f52cb-78d6-11ea-bc32-1e00400108b3', jobstatus : 2, jobid : u'8806aed0-896c-4264-801e-d570217f4ae0', jobresultcode : 530, jobinstanceid : u'4cdc1214-942c-436c-b6ae-738f3220b1cf', jobresulttype : u'object', jobinstancetype : u'Snapshot', completed : u'2020-04-07T16:46:13+0000', accountid : u'964f4d54-78d6-11ea-bc32-1e00400108b3'}\n"]

VMware tests that failed:

'org.apache.cloudstack.api.command.admin.vm.MigrateVMCmd\', userid : u\'3d686233-78da-11ea-8ab1-1e009101089c\', jobstatus : 2, jobid : u\'aace648a-e269-49f8-982b-3c07bf5baa11\', jobresultcode : 530, jobresulttype : u\'object\', jobresult : {errorcode : 530, errortext : u"MigrationCommand failed due to Exception: java.lang.RuntimeException\\nMessage: **The target host does not support the virtual machine\'s current hardware requirements.**\\n"}, accountid : u\'3d685acc-78da-11ea-8ab1-1e009101089c\'}\n']
That is funny...

--- a/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/DefaultSnapshotStrategy.java
+++ b/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/DefaultSnapshotStrategy.java
@@ -288,10 +288,12 @@ public class DefaultSnapshotStrategy extends SnapshotStrategyBase {
             // This may occur for instance when they are created only on primary because
             // snapshot.backup.to.secondary was set to false.
             if (snapshotOnPrimaryInfo == null) {
-                s_logger.debug("Snapshot id:" + snapshotId + " not found on primary storage, skipping snapshot deletion on primary and marking it destroyed");
+                if (s_logger.isDebugEnabled()) {
+                    s_logger.debug(String.format("Snapshot (id: %d) not found on primary storage, skipping snapshot deletion on primary and marking it destroyed", snapshotId));
+                }
                 snapshotVO.setState(Snapshot.State.Destroyed);
                 snapshotDao.update(snapshotId, snapshotVO);
-                return true;
+                deletedOnPrimary = true;
             } else {
                 SnapshotObject obj = (SnapshotObject) snapshotOnPrimaryInfo;
                 try {
@@ -365,15 +367,21 @@ public class DefaultSnapshotStrategy extends SnapshotStrategyBase {
     private boolean deleteSnapshotOnPrimary(Long snapshotId, SnapshotInfo snapshotOnPrimaryInfo) {
         SnapshotDataStoreVO snapshotOnPrimary = snapshotStoreDao.findBySnapshot(snapshotId, DataStoreRole.Primary);
         if (isSnapshotOnPrimaryStorage(snapshotId)) {
-            s_logger.debug("Snapshot reference is found on primary storage for snapshot id:" + snapshotId + ", performing snapshot deletion on primary");
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug(String.format("Snapshot reference is found on primary storage for snapshot id: %d, performing snapshot deletion on primary", snapshotId));
+            }
             if (snapshotSvr.deleteSnapshot(snapshotOnPrimaryInfo)) {
                 snapshotOnPrimary.setState(State.Destroyed);
                 snapshotStoreDao.update(snapshotOnPrimary.getId(), snapshotOnPrimary);
-                s_logger.debug("Successfully deleted snapshot id:" + snapshotId + " on primary storage");
+                if (s_logger.isDebugEnabled()) {
+                    s_logger.debug(String.format("Successfully deleted snapshot id: %d on primary storage", snapshotId));
+                }
                 return true;
             }
         } else {
-            s_logger.debug("Snapshot reference is not found on primary storage for snapshot id:" + snapshotId + ", skipping snapshot deletion on primary");
+            if (s_logger.isDebugEnabled()) {
+                s_logger.debug(String.format("Snapshot reference is not found on primary storage for snapshot id: %d, skipping snapshot deletion on primary", snapshotId));
+            }
             return true;
         }
         return false;

--- a/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/DefaultSnapshotStrategy.java
+++ b/engine/storage/snapshot/src/main/java/org/apache/cloudstack/storage/snapshot/DefaultSnapshotStrategy.java
@@ -287,18 +287,25 @@ public class DefaultSnapshotStrategy extends SnapshotStrategyBase {
             // Here we handle snapshots which are to be deleted but are not marked as destroyed yet.
             // This may occur for instance when they are created only on primary because
             // snapshot.backup.to.secondary was set to false.
-            SnapshotObject obj = (SnapshotObject) snapshotOnPrimaryInfo;
-            try {
-                obj.processEvent(Snapshot.Event.DestroyRequested);
-                deletedOnPrimary = deleteSnapshotOnPrimary(snapshotId, snapshotOnPrimaryInfo);
-                if (!deletedOnPrimary) {
-                    obj.processEvent(Snapshot.Event.OperationFailed);
-                } else {
-                    obj.processEvent(Snapshot.Event.OperationSucceeded);
+            if (snapshotOnPrimaryInfo == null) {
+                s_logger.debug("Snapshot id:" + snapshotId + " not found on primary storage, skipping snapshot deletion on primary and marking it destroyed");
+                snapshotVO.setState(Snapshot.State.Destroyed);
+                snapshotDao.update(snapshotId, snapshotVO);
+                return true;
+            } else {
+                SnapshotObject obj = (SnapshotObject) snapshotOnPrimaryInfo;
+                try {
+                    obj.processEvent(Snapshot.Event.DestroyRequested);
+                    deletedOnPrimary = deleteSnapshotOnPrimary(snapshotId, snapshotOnPrimaryInfo);
+                    if (!deletedOnPrimary) {
+                        obj.processEvent(Snapshot.Event.OperationFailed);
+                    } else {
+                        obj.processEvent(Snapshot.Event.OperationSucceeded);
+                    }
+                } catch (NoTransitionException e) {
+                    s_logger.debug("Failed to set the state to destroying: ", e);
+                    deletedOnPrimary = false;
                 }
-            } catch (NoTransitionException e) {
-                s_logger.debug("Failed to set the state to destroying: ", e);
-                deletedOnPrimary = false;
             }
         }
         return deletedOnPrimary;
@@ -352,14 +359,21 @@ public class DefaultSnapshotStrategy extends SnapshotStrategyBase {
     }
 
     /**
-     * Deletes the snapshot on primary storage. It can return false when the snapshot was not stored on primary storage; however this does not means that it failed to delete the snapshot. </br>
-     * In case of failure, it will throw one of the following exceptions: CloudRuntimeException, InterruptedException, or ExecutionException. </br>
+     * Deletes the snapshot on primary storage. It returns true when the snapshot was not found on primary storage; </br>
+     * In case of failure while deleting the snapshot, it will throw one of the following exceptions: CloudRuntimeException, InterruptedException, or ExecutionException. </br>
      */
     private boolean deleteSnapshotOnPrimary(Long snapshotId, SnapshotInfo snapshotOnPrimaryInfo) {
         SnapshotDataStoreVO snapshotOnPrimary = snapshotStoreDao.findBySnapshot(snapshotId, DataStoreRole.Primary);
-        if (isSnapshotOnPrimaryStorage(snapshotId) && snapshotSvr.deleteSnapshot(snapshotOnPrimaryInfo)) {
-            snapshotOnPrimary.setState(State.Destroyed);
-            snapshotStoreDao.update(snapshotOnPrimary.getId(), snapshotOnPrimary);
+        if (isSnapshotOnPrimaryStorage(snapshotId)) {
+            s_logger.debug("Snapshot reference is found on primary storage for snapshot id:" + snapshotId + ", performing snapshot deletion on primary");
+            if (snapshotSvr.deleteSnapshot(snapshotOnPrimaryInfo)) {
+                snapshotOnPrimary.setState(State.Destroyed);
+                snapshotStoreDao.update(snapshotOnPrimary.getId(), snapshotOnPrimary);
+                s_logger.debug("Successfully deleted snapshot id:" + snapshotId + " on primary storage");
+                return true;
+            }
+        } else {
+            s_logger.debug("Snapshot reference is not found on primary storage for snapshot id:" + snapshotId + ", skipping snapshot deletion on primary");
             return true;
         }
         return false;

